modules/module-postgres-storage/src/storage/batch/PostgresBucketBatch.ts

```diff
@@ -0,0 +1,888 @@
+import * as lib_postgres from '@powersync/lib-service-postgres';
+import { container, DisposableObserver, errors, logger } from '@powersync/lib-services-framework';
+import { storage, utils } from '@powersync/service-core';
+import * as sync_rules from '@powersync/service-sync-rules';
+import * as timers from 'timers/promises';
+import * as t from 'ts-codec';
+import { CurrentBucket, CurrentData, CurrentDataDecoded } from '../../types/models/CurrentData.js';
+import { models, RequiredOperationBatchLimits } from '../../types/types.js';
+import { NOTIFICATION_CHANNEL, sql } from '../../utils/db.js';
+import { pick } from '../../utils/ts-codec.js';
+import { batchCreateCustomWriteCheckpoints } from '../checkpoints/PostgresWriteCheckpointAPI.js';
+import { cacheKey, encodedCacheKey, OperationBatch, RecordOperation } from './OperationBatch.js';
+import { PostgresPersistedBatch } from './PostgresPersistedBatch.js';
+
+export interface PostgresBucketBatchOptions {
+  db: lib_postgres.DatabaseClient;
+  sync_rules: sync_rules.SqlSyncRules;
+  group_id: number;
+  slot_name: string;
+  last_checkpoint_lsn: string | null;
+  no_checkpoint_before_lsn: string;
+  store_current_data: boolean;
+  keep_alive_op?: string | null;
+  /**
+   * Set to true for initial replication.
+   */
+  skip_existing_rows: boolean;
+  batch_limits: RequiredOperationBatchLimits;
+}
+
+/**
+ * Intermediate type which helps for only watching the active sync rules
+ * via the Postgres NOTIFY protocol.
+ */
+const StatefulCheckpoint = models.ActiveCheckpoint.and(t.object({ state: t.Enum(storage.SyncRuleState) }));
+type StatefulCheckpointDecoded = t.Decoded<typeof StatefulCheckpoint>;
+
+// The limits here are not as strict as MongoDB
+const MAX_ROW_SIZE = 40_000_000;
```

Comment:
I'd recommend keeping the same limit as we have for MongoDB for the initial release, then re-evaluate what we'd like here - this can potentially have big impacts on memory usage in many parts of the service. And it's easy to increase this later, but we can't decrease it without breaking backwards-compatibility.

Reply:
👍  Updated the value now.